uarray/_uarray_dispatch.cxx

```diff
@@ -917,6 +1173,89 @@ PyObject * Function::get_replacer(Function * self) {
+    0,                                 /* tp_itemsize */
+    (destructor)BackendState::dealloc, /* tp_dealloc */
+    0,                                 /* tp_print */
+    0,                                 /* tp_getattr */
+    0,                                 /* tp_setattr */
+    0,                                 /* tp_reserved */
+    0,                                 /* tp_repr */
+    0,                                 /* tp_as_number */
+    0,                                 /* tp_as_sequence */
+    0,                                 /* tp_as_mapping */
+    0,                                 /* tp_hash  */
+    0,                                 /* tp_call */
+    0,                                 /* tp_str */
+    0,                                 /* tp_getattro */
+    0,                                 /* tp_setattro */
+    0,                                 /* tp_as_buffer */
+    Py_TPFLAGS_DEFAULT,                /* tp_flags */
+    0,                                 /* tp_doc */
+    0,                                 /* tp_traverse */
+    0,                                 /* tp_clear */
+    0,                                 /* tp_richcompare */
+    0,                                 /* tp_weaklistoffset */
+    0,                                 /* tp_iter */
+    0,                                 /* tp_iternext */
+    BackendState_Methods,              /* tp_methods */
+    0,                                 /* tp_members */
+    0,                                 /* tp_getset */
+    0,                                 /* tp_base */
+    0,                                 /* tp_dict */
+    0,                                 /* tp_descr_get */
+    0,                                 /* tp_descr_set */
+    0,                                 /* tp_dictoffset */
+    0,                                 /* tp_init */
+    0,                                 /* tp_alloc */
+    BackendState::new_,                /* tp_new */
+};
+
+PyObject * get_state(PyObject * /* self */, PyObject * /* args */) {
+  py_ref new_tuple = py_ref::steal(PyTuple_New(0));
+  if (!new_tuple)
+    return nullptr;
+
+  py_ref ref = py_ref::steal(PyObject_Call(
+      reinterpret_cast<PyObject *>(&BackendStateType), new_tuple.get(), NULL));
+  BackendState * output = reinterpret_cast<BackendState *>(ref.get());
+
+  output->locals = local_state;
+  output->globals = global_state;
+
+  return ref.release();
+}
+
+PyObject * set_state(PyObject * /* self */, PyObject * args) {
+  PyObject * arg;
+  if (!PyArg_ParseTuple(args, "O", &arg))
+    return nullptr;
+
+  if (!PyObject_IsInstance(
+          arg, reinterpret_cast<PyObject *>(&BackendStateType))) {
+    PyErr_SetString(
+        PyExc_TypeError, "state must be a uarray._BackendState object.");
+    return nullptr;
+  }
+
+  BackendState * state = reinterpret_cast<BackendState *>(arg);
+  local_state = state->locals;
+  global_state = state->globals;
```

Comment:
It's occurred to me that the main motivation for this was multithreading, yet this is completely thread unsafe. We are modifying the global state without locks and it's visible to all threads.

One remedy would be to only expose the thread's local state. That should be sufficient for multithreading since the global state is shared anyway. However, it still doesn't solve `dask` because we *do* rely on different global state due to multiprocessing and delayed execution.

A more complex solution might be to change the global state to be a `thread_local std::shared_ptr<global_domain_map_t>`. By default they all point to the same global state but inside of a `ua.set_state` block, the thread has its own separate context.

Perhaps what this really means is that registered backbends are a bad idea.

Reply:
How about simply making `global_state` thread-local as well?

Reply:
I suppose that would work but then they aren't really global any longer.

Reply:
locals are local to a context, and globals are local to a thread.

Reply:
Okay, should probably add a note in the docs then.

Reply:
Oh wait, this breaks SciPy as well... We use `set_global_backend` to ensure `scipy.fft.fft` defaults to the `pocketfft` implementation.

Reply:
That happens at module import time which is a one time operation, not once per thread.

Reply:
One thing we can do is just note that global registration should only happen at import.

Reply:
That would still be an issue for imports inside function scope.

Reply:
> By default they all point to the same global state but inside of a `ua.set_state` block, the thread has its own separate context.

I did this with a bit less churn... Hopefully it's correct.